New Post: GUI Challenges - Tabs (#4668)
GoogleChrome/web.dev: src/site/content/en/blog/building-a-tabs-component/index.md

title: Building a Tabs component
subhead: A foundational overview of how to build a tabs component similar to those found in iOS and Android apps.
authors:
  - adamargyle
description: A foundational overview of how to build a tabs component similar to those found in iOS and Android apps.
date: 2021-02-17
hero: hero.png
thumbnail: thumb.png
tags:
  - blog
  - css
  - dom
  - javascript
  - layout
  - mobile
  - ux
---

In this post I want to share thinking on building a Tabs component for the web
that is responsive, supports multiple device inputs, and works across browsers.
Try the [demo](https://gui-challenges.web.app/tabs/dist/).

<figure class="w-figure w-figure--fullbleed">
  <video playsinline controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/atoms-sandbox.google.com.a.appspot.com/gui-tabs/mouse-and-keyboard.mp4">
  </video>
  <figcaption class="w-figure">
    <a href="https://gui-challenges.web.app/tabs/dist/">Demo</a>
  </figcaption>
</figure>

If you prefer video, here's a YouTube version of this post:

{% YouTube 'PzvdREGR0Xw' %}

## Overview

Tabs are a common component of design systems but can take many shapes and

> forms. First there were desktop tabs built on `<frames>`, and now we have

Reviewer: ```suggestion
forms. First there were desktop tabs built on `<frame>` element, and now we have
```

buttery mobile components that animate content based on physics properties.
They're all trying to do the same thing: save space. 

Today, the essentials of a tabs user experience is a button navigation area
which toggles the visibility of content in a display frame. Many different
content areas share the same space, but are conditionally presented based on the
button selected in the navigation.

<figure class="w-figure">
  <img class="w-screenshot" src="./tabs-variety.png" alt="">
  <figcaption class="w-figcaption">
    A collage of tab component web design styles from over the past 10 years
  </figcaption>
</figure>

## Web Tactics

All in all I found this component pretty straightforward to build, thanks to a
few critical web platform features:
- `scroll-snap-points` for elegant swipe and keyboard interactions with
  appropriate scroll stop positions
- [Deep links](https://en.wikipedia.org/wiki/Deep_linking) via URL hashes for
  browser handled in-page scroll anchoring and sharing support
- Screen reader support with `<a>` and `id="#hash"` element markup
- `prefers-reduced-motion` for enabling crossfade transitions and instant
  in-page scrolling

> - The in draft `@scroll-timeline` web feature for dynamically underlining and

Reviewer: ```suggestion
- The in-draft `@scroll-timeline` web feature for dynamically underlining and
```

    <article></article>
    <article></article>
  </section>
</snap-tabs>
```

I can establish connections between the `<a>` and `<article>` elements with
`href` and `id` properties like this:

```html/3,10
<snap-tabs>
  <header>
    <nav>
      <a href="#responsive"></a>
      <a href="#accessible"></a>
      <a href="#overscroll"></a>
      <a href="#more"></a>
    </nav>
  </header>
  <section>
    <article id="responsive"></article>
    <article id="accessible"></article>
    <article id="overscroll"></article>
    <article id="more"></article>
  </section>
</snap-tabs>
```

I next filled the articles with mixed amounts of lorem, and the links with a
mixed length and image set of titles. With content to work with, we can begin
layout.

### Scrolling layouts {: #overscroll }

There are 3 different types of scroll areas in this component:
- The navigation <b style="color: #FF00E2;">(pink)</b> is horizontally
  scrollable
- The content area <b style="color: #008CFF;">(blue)</b> is horizontally
  scrollable
- Each article item <b style="color: #2FD800;">(green)</b> is vertically
  scrollable.

<figure class="w-figure">
  <img class="w-screenshot" src="./3-scrollers.png" alt="">

> </figure>

Reviewer: Thanks for the bigger arrows; much easier to see now

  </header>
  <section></section>
</snap-tabs>
```
{% endCompare %}

{% Compare 'better', 'CSS' %}

```css/1-2
header {
  display: flex;
  flex-direction: column;
}
```
{% endCompare %}
</div>

The `.snap-indicator` should travel horizontally with the group of links, and
this header layout helps set that stage. No absolute positioned elements here! 

<figure class="w-figure">
  <img class="w-screenshot" src="./header-layout.png" alt="">
</figure>

Next, the scroll styles. It turns out that we can share the scroll styles
between our 2 horizontal scroll areas (header and section), so I made a utility
class, `.scroll-snap-x`. 

```css
.scroll-snap-x {
  /* browser decide if x is ok to scroll and show bars on, y hidden */
  overflow: auto hidden;
  /* prevent scroll chaining on x scroll */
  overscroll-behavior-x: contain;
  /* scrolling should snap children on x */
  scroll-snap-type: x mandatory;
  
  @media (hover: none) {
    scrollbar-width: none;

    &::-webkit-scrollbar {
      width: 0;
      height: 0;
    }
  }
}
```

Each needs overflow on the x axis, scroll containment to trap overscroll, hidden
scrollbars for touch devices and lastly scroll-snap for locking content

> presentation areas. Our keyboard tab order is nice and any interactions guide

Reviewer: ```suggestion
presentation areas. Our keyboard tab order is accessible and any interactions guide
```

</figure>

I think seeing the scrollbar gutter in this component helps clearly show where
the scroll areas are, the direction they support, and how they interact with
each other. Consider how each of these scroll window frames also are flex or
grid parents to a layout.

DevTools can help us visualize this:

<figure class="w-figure">
  <img class="w-screenshot" src="./3-scroll-areas.png" alt="">
  <figcaption class="w-figure">
    Chromium Devtools, showing the flexbox nav element layout full of anchor elements, 
    the grid section layout full of article elements, and the article 
    elements full of paragraphs and a heading element.
  </figcaption>
</figure>

The scroll layouts are complete: snapping, deep linkable, and keyboard
accessible. Strong foundation for UX enhancements, style and delight.

#### Feature highlight

Scroll snapped children maintain their locked position during resize. This means
JavaScript won't need to bring anything into view on device rotate or browser
resize. Try it out in Chromium DevTools [Device
Mode](https://developers.google.com/web/tools/chrome-devtools/device-mode) by

> selecting any mode other than Responsive, and then resizing the device frame.

Reviewer: ```suggestion
selecting any mode other than **Responsive**, and then resizing the device frame.
```

different transition styles. I had a ton of fun building this.

<figure class="w-figure">
  <video playsinline controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/atoms-sandbox.google.com.a.appspot.com/gui-tabs/prefer-reduced-motion-toggle-mini.mp4">
  </video>
</figure>

```css
@media (prefers-reduced-motion: reduce) {
  snap-tabs > header a {
    border-block-end: var(--indicator-size) solid hsl(var(--accent) / 0%);
    transition: color .7s ease, border-color .5s ease;

    &:is(:target,:active,[active]) {
      color: var(--text-active-color);
      border-block-end-color: hsl(var(--accent));
    }
  }

  snap-tabs .snap-indicator {
    visibility: hidden;
  }
}
```

I hide the `.snap-indicator` when the user prefers reduced motion since I don't
need it anymore. Then I replace it with `border-block-end` styles and a
`transition`. Also notice in the tabs interaction that the active nav item not
only has a brand underline highlight, but it's text color also is darker. The
active element has higher text color contrast and a bright underlight accent.

Just a few extra lines of CSS will make someone feel seen (in the sense that

> we're thoughtfully respecting their motion preferences), love it.

Reviewer: ```suggestion
we're thoughtfully respecting their motion preferences). I love it.
```

position of the indicator change from one section to the next, tracking
perfectly with scroll.

You may not have noticed, but I'm very proud of the transition of color as the
highlighted navigation item becomes selected.

<figure class="w-figure">
  <video playsinline controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/atoms-sandbox.google.com.a.appspot.com/gui-tabs/scroll-color.mp4">
  </video>
</figure>

The unselected lighter grey appears even more pushed back when the highlighted
item has more contrast. It's common to transition color for text, like on hover
and when selected, but it's next-level to transition that color on scroll,
synchronized with the underline indicator. 

Here's how I did it:

```js
tabnavitems.forEach(navitem => {
  navitem.animate({
      color: [...tabnavitems].map(item => 
        item === navitem
          ? `var(--text-active-color)`
          : `var(--text-color)`)
    }, {
      duration: 1000,
      fill: 'both',
      timeline: sectionScrollTimeline,
    }
  );
});
```

Each tab nav link needs this new color animation, tracking the same scroll

> timeline as the underline indicator. I use the same timeline as before:, since

Reviewer: ```suggestion
timeline as the underline indicator. I use the same timeline as before: since
```

animation we want. As I did before, I create 4 keyframes in the loop, and return
colors.

```js
[...tabnavitems].map(item => 
  item === navitem
    ? `var(--text-active-color)`
    : `var(--text-color)`)

// results in 4 array items, which represent 4 keyframe states
// [
  "var(--text-active-color)", 
  "var(--text-color)", 
  "var(--text-color)", 
  "var(--text-color)",
]
```

The keyframe with the color `var(--text-active-color)` highlights the link, and
it's otherwise a standard text color. The nested loop there makes it relatively
straightforward, as the outer loop is each nav item, and the inner loop is each
navitem's personal keyframes. I check if the outer loop element is the same as
the inner loop one, and use that to know when it's selected.

I had a lot of fun writing this. So much.

### Even more Javascript enhancements {: #js }

It's worth a reminder that the core of what I'm showing you here works without
JavaScript. With that said, let's see how we can enhance it when JS is
available.

#### Deep links 

Deep links are more of a mobile term, but I think the intent of the deep link is
met here with tabs in that you can share a URL directly to a tab's contents. The

> browser will in-page navigate to the ID that is matched in the URL hash.I found

Reviewer: ```suggestion
browser will in-page navigate to the ID that is matched in the URL hash. I found
```

this `onload` handler made the effect across platforms.

```js
window.onload = () => {
  if (location.hash) {
    tabsection.scrollLeft = document
      .querySelector(location.hash)
      .offsetLeft;
  }
}
```

#### Scroll end synchronization 

Our users aren't always clicking or using a keyboard, sometimes they're just
free scrolling, as they should be able to. When the section scroller stops

> scrolling, wherever it lands needs to be matched in the top navigation bar.
> Here's how I wait for scroll end.
> 
> <figure class="w-figure">
>   <video playsinline controls autoplay loop muted class="w-screenshot">
>     <source src="https://storage.googleapis.com/atoms-sandbox.google.com.a.appspot.com/gui-tabs/scrollend-sync.mp4">
>   </video>
> </figure>
> 

Reviewer: ```suggestion
scrolling, wherever it lands needs to be matched in the top navigation bar.

<figure class="w-figure">
  <video playsinline controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/atoms-sandbox.google.com.a.appspot.com/gui-tabs/scrollend-sync.mp4">
  </video>
</figure>

Here's how I wait for scroll end:
```
Moving that last sentence to be next to the code it's referring to
